p4-16/psa/PSA.mdk

```diff
@@ -172,18 +172,21 @@ backend is responsible for mapping the PRE extern APIs to the
 appropriate mechanisms for performing these operations in the
 hardware.
 
-The PRE is instantiated by the architecture and a P4 program can use
-it directly. It is an error to instantiate the PRE multiple times.
+The PRE extern object has no constructor, and thus it cannot be
+instantiated in the user's P4 program. The architecture instantiates
+it exactly once, without requiring the user's P4 program to
+instantiate it.
 The PRE is made available to the Ingress programmable block using the
 same mechanism as packet_in. A corresponding Buffering and Queuing
 Engine (BQE) extern is defined for the Egress pipeline (@see
 BufferingQueueingEngine).
 
-Note: some of these operations may not be implemented as primitive
-operations on a certain target. However, All of the operations can be
-implemented as a combination of other operations. Applications that
-rely on non-primitive operations may incur significant performance
-penalties, however, they should be functionally correct.
+**Note to implementers**: some of these operations may not be
+implemented as primitive operations on a certain target. However, the
+goal is that all operations should be implementable as a combination
+of other operations. Target documentation should highlight such
```

Comment:
@jafingerhut the architecture is the time to talk about hardware mapping. These are not the operations in the p4 program. Primitive operations and other operations are referring to what the hardware target supports natively or not. I will add clarifications to that effect. 

I think we should give users a hint that performance is target dependent -- it will be better to be warned than be surprised by performance variability. If you have any suggestions on how to express this in a more marketable way, by all means!